File: lru_sim/lru_sim.py
from circularLinkedList import CircularLinkedList


class CacheSimulator:
    def __init__(self, cache_slots):
        self.cache_slots = cache_slots
        self.cache_hit = 0
        self.tot_cnt = 1
        self.container = CircularLinkedList()
    
    def do_sim(self, page):
        self.tot_cnt += 1
        if self.container.index(page) != -2:
            self.cache_hit += 1
            self.container.remove(page)
            self.container.append(page)
            return
        if self.container.size() < self.cache_slots:
            self.container.append(page)
        else:
            self.container.pop(0)
            self.container.append(page)

        # if data가 겹침:
        #         cache_hit += 1
        #         (append 안하는데) 자리는 recently, 즉 tail쪽으로 이동 -> update

        # if cache_slot < cache_slots: 캐시 슬롯에 빈자리 있음
        #     그냥 append # append 구문을 위로 빼도 되지만 실제 상황에서 어려울 수 있다 (메모리 이슈)
        
        # else: 캐시 슬롯에 빈자리 없음
        #     기존 head를 쫓아내고 new tail로 새 page append

        
    def print_stats(self):
        print("cache_slot = ", self.cache_slots, "cache_hit = ", self.cache_hit, "hit ratio = ", self.cache_hit / self.tot_cnt)


if __name__ == "__main__":

    data_file = open("./linkbench.trc")
    lines = data_file.readlines()
    for cache_slots in range(100, 1001, 100):
        cache_sim = CacheSimulator(cache_slots)
        for line in lines:
            page = line.split()[0] # "\n" escape sequence 삭제
            cache_sim.do_sim(page)
        
        cache_sim.print_stats()

File: lru_sim/lru_sim.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include "circularLinkedList.h"

#define INITIAL_CAPACITY 10000
#define INCREMENT_SIZE 10000

typedef struct CacheSimulator {
    int cache_slots;
    int cache_hit;
    int tot_cnt;
    CircularLinkedList container;
} CacheSimulator;

void do_sim(CacheSimulator* sim, char* page) {
    sim->tot_cnt++;
    if (index(&(sim->container), page) != -2) {
        sim->cache_hit++;
        // printf("%s\n", page);
        removeItem(&(sim->container), page);
        append(&(sim->container), page);
        return;
    }
    append(&(sim->container), page);
    if (size(&(sim->container)) >= sim->cache_slots)
        pop(&(sim->container));
}

void print_stats(CacheSimulator* sim) {
    printf("cache slots = %d ", sim->cache_slots);
    printf("cache hits = %d ", sim->cache_hit);
    printf("hit ratio = %.5f\n", (float)sim->cache_hit / sim->tot_cnt);
}

int main() {
    FILE* data_file = fopen("./linkbench.trc", "r");
    if (data_file == NULL) {
        perror("Error opening file");
        return 1;
    }

    char line[256];
    int capacity = INITIAL_CAPACITY;
    char** lines = (char**)malloc(capacity * sizeof(char*));
    int lines_count = 0;

    while (fgets(line, sizeof(line), data_file) != NULL) {
        line[strcspn(line, "\n")] = 0;
        if (lines_count >= capacity) {
            capacity += INCREMENT_SIZE;
            lines = (char**)realloc(lines, capacity * sizeof(char*));
            if (lines == NULL) {
                perror("Error reallocating memory");
                return 1;
            }
        }
        lines[lines_count] = strdup(strtok(line, " "));
        lines_count++;
    }

    fclose(data_file);

    CacheSimulator* cache_sim;
    cache_sim = (CacheSimulator*)malloc(sizeof(CacheSimulator));
    if (cache_sim == NULL) {
        perror("Error allocating memory");
        return 1;
    }
    // printf("%d\n", lines_count);
    for (int cache_slots = 100; cache_slots <= 1000; cache_slots += 100) {
        cache_sim->cache_slots = cache_slots;
        cache_sim->cache_hit = 0;
        cache_sim->tot_cnt = 1;
        initCircularLinkedList(&(cache_sim->container));
        for (int i = 0; i < lines_count; ++i) {
        // printf("%d\n", size(&(cache_sim->container)));
            do_sim(cache_sim, lines[i]);
        }

        print_stats(cache_sim);
    }

    free(cache_sim); // CacheSimulator 메모리 해제
    free(lines);

    return 0;
}
